packages/muelu/src/MueCentral/MueLu_HierarchyUtils_def.hpp

```diff
@@ -98,15 +98,52 @@ namespace MueLu {
                                      !IsParamMuemexVariable(name), Exceptions::InvalidArgument,
                                      std::string("MueLu::Utils::AddNonSerializableDataToHierarchy: parameter list contains unknown data type(") + name + ")");
 
+          // Get a valid communicator and lib
+          RCP<const Teuchos::Comm<int> > comm;
+          if (!level->GetComm().is_null())
+            comm = level->GetComm();
+          else if (level->IsAvailable("A")) {
+            RCP<Matrix> mat;
+            level->Get("A", mat);
+            comm = mat->getMap()->getComm();
+          } else {
+            RCP<Level> level0 = H.GetLevel(0);
+            if (!level0->GetComm().is_null())
+              comm = level0->GetComm();
+            else {
+              RCP<Matrix> mat;
+              level0->Get("A", mat);
+              comm = mat->getMap()->getComm();
+            }
+          }
+          Xpetra::UnderlyingLib lib = level->lib();
+
           if (name == "A") {
-            level->Set(name, Teuchos::getValue<RCP<Matrix > > (levelListEntry->second),NoFactory::get());
+            RCP<Matrix> mat;
+            if (levelListEntry->second.isType<std::string>())
+              // We might also want to read maps here.
+              mat = Xpetra::IO<Scalar,LocalOrdinal,GlobalOrdinal,Node>::Read(Teuchos::getValue<std::string>(levelListEntry->second), lib, comm);
+            else
+              mat = Teuchos::getValue<RCP<Matrix > > (levelListEntry->second);
+            level->Set(name, mat, NoFactory::get());
             M->SetFactory(name, NoFactory::getRCP()); // TAW: not sure about this: be aware that this affects all levels
                                                       //      However, A is accessible through NoFactory anyway, so it should
                                                       //      be fine here.
           }
           else if(name == "P" || name == "R" || name == "K" || name == "M") {
```

Comment:
What are "K" and "M"?

Reply:
I'm assuming something like mass and stiffness, but no idea where in MueLu that's actually used.